[WIP] [DOCS] Add Nebari installation guides and getting-started as first directions (#45)
nebari-dev/nebari: docs/docs/how-tos/nebari-aws.md

---
id: nebari-aws
title: How to deploy Nebari on AWS
---

# How to deploy Nebari on AWS

> A basic overview to deploy Nebari on AWS

Reviewer: ```suggestion
A basic overview of how to deploy Nebari on AWS
```

> This guide is to help first time users set up an Amazon Web Services (AWS) account specifically with the intention to use and deploy Nebari at a production scale. In this guide we will walk you through the following steps:

Reviewer: ```suggestion
This guide is to help first-time users set up an Amazon Web Services (AWS) account specifically for the purpose of using and deploying Nebari at a production scale. In this guide we will walk you through the following steps:
```

> - [Sign up for Amazon Web Services](#sign-up-for-google-cloud-platform);

Reviewer: Fix internal doc ref link to AWS from GCP

- [Set up IAM user and access keys](#authentication);
- [Initialize Nebari with your IAM user credentials](#nebari-initialize);
- [Deploy Nebari](#deploying-nebari)

For those already familiar to AWS services, feel free to skip this first step and jump straight to the [Nebari authentication](#authentication) section of this guide.

> ## Sing up for Amazon Web Services

Reviewer: ```suggestion
## Sign up for Amazon Web Services
```

> This documentation assumes that the user is already familiar with [AWS Identity and Access Management (IAM)](https://docs.aws.amazon.com/IAM/latest/UserGuide/introduction.html) and has a prior knowledge regarding [AWS billing and cost usage](https://aws.amazon.com/eks/pricing/) for Kubernetes related services.

Reviewer: ```suggestion
This documentation assumes that you are already familiar with [AWS Identity and Access Management (IAM)](https://docs.aws.amazon.com/IAM/latest/UserGuide/introduction.html) and that you have prior knowledge regarding [AWS billing and cost usage](https://aws.amazon.com/eks/pricing/) for Kubernetes related services.
```

If you are new to AWS, we advise you to first [sign up for a free account](https://aws.amazon.com/free/free-tier/) to get a better understanding of the platform and its features. Please refer to [Amazon VPC (Virtual Private Cloud)](https://aws.amazon.com/vpc/?nc1=h_ls) and [Amazon EKS Prerequisites](https://docs.aws.amazon.com/eks/latest/userguide/getting-started-console.html#eks-prereqs) for more information on account types and prerequisites for managing Kubernetes clusters.

:::warning Warning

> The Nebari deployment on AWS will **NOT** fall into `free tier` usage as some of its inner components will lead to [special charges](https://aws.amazon.com/eks/pricing/). Therefore, we recommend that you check [AWS pricing documentation](https://aws.amazon.com/ec2/pricing/) or contact your cloud administrator for more information. If you provision resources outside of the free tier, you may be charged. We're not responsible for any charges that may incur.

Reviewer: ```suggestion
A Nebari deployment on AWS will **NOT** fall into `free tier` usage as some of its inner components will lead to [special charges](https://aws.amazon.com/eks/pricing/). Therefore, we recommend that you check [AWS pricing documentation](https://aws.amazon.com/ec2/pricing/) or contact your cloud administrator for more information. If you provision resources outside of the free tier, you may be charged. We're not responsible for any charges that may incur.
```

I think 'a' would be more commonly used than 'the' here.

> The Nebari deployment on AWS will **NOT** fall into `free tier` usage as some of its inner components will lead to [special charges](https://aws.amazon.com/eks/pricing/). Therefore, we recommend that you check [AWS pricing documentation](https://aws.amazon.com/ec2/pricing/) or contact your cloud administrator for more information. If you provision resources outside of the free tier, you may be charged. We're not responsible for any charges that may incur.

Reviewer: ```suggestion
A Nebari deployment on AWS will **NOT** fall into `free tier` usage as some of its inner components will lead to [special charges](https://aws.amazon.com/eks/pricing/). Therefore, we recommend that you check [AWS pricing documentation](https://aws.amazon.com/ec2/pricing/) or contact your cloud administrator for more information. If you provision resources outside of the free tier, you may be charged. We're not responsible for any charges you may incur if this happens.
```

I think 'a' would be more commonly used than 'the', here.

:::

## Authentication

> In order to Nebari make requests against the aws API and create it's infrastructure, an authentication method with the appropriate permissions will be required. The easiest way to do this is using a [IAM user](https://docs.aws.amazon.com/IAM/latest/UserGuide/introduction.html) to allow access over your AWS account and EKS cluster management.

Reviewer: ```suggestion
In order for Nebari to make requests against the AWS API and create its infrastructure, an authentication method with the appropriate permissions will be required. The easiest way to do this is using an [IAM user](https://docs.aws.amazon.com/IAM/latest/UserGuide/introduction.html) with suitable permissions for your AWS account and EKS cluster management.
```

> As a [best practice](https://docs.aws.amazon.com/IAM/latest/UserGuide/best-practices.html#lock-away-credentials), do not use the AWS account root user for any task where it's not required. Instead, create a new IAM user for each person that requires administrator access. Then make those users administrators by placing the users into an "Administrators" user group to which you attach the `AdministratorAccess` managed policy.

Reviewer: ```suggestion
As a [best practice](https://docs.aws.amazon.com/IAM/latest/UserGuide/best-practices.html#lock-away-credentials), do not use the AWS account root user for any task where it's not required. Instead, create a new IAM user for each person that requires administrator access. Then make those users administrators by placing them into an "Administrators" user group, to which you attach the `AdministratorAccess` managed policy.
```

> If its your first time creating a new IAM user, please refer to [Creating your first IAM admin user and user group](https://docs.aws.amazon.com/IAM/latest/UserGuide/getting-started_create-admin-group.html) for more information. Otherwise, see [Creating an IAM user in your AWS account](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_users_create.html) for more information on how to create a IAM user.

Reviewer: ```suggestion
If it's your first time creating a new IAM user, please refer to [Creating your first IAM admin user and user group](https://docs.aws.amazon.com/IAM/latest/UserGuide/getting-started_create-admin-group.html) for more information. Otherwise, see [Creating an IAM user in your AWS account](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_users_create.html) for more information on how to create an IAM user.
```

Follow these steps to set up your access keys and user accounts:

- Go to [IAM in your AWS Console](https://console.aws.amazon.com/iam/home), then **Users**, and **Add Users**;

> - Give the user a name, and tick **Access Key - Programmatic access**

Reviewer: Check capitalization - is it "Programmatic Access", capital A?

- Click **Next**

> ![alt text for screen readers](/img/how-tos-aws-new-iam-user.png "Text to show on mouseover")

Reviewer: Needs alt text and mouseover text

- Select **Attach existing policies directly**, then select `AdministratorAccess` from the list of policies. For more information, please refer to [Policies and permissions in IAM](https://docs.aws.amazon.com/IAM/latest/UserGuide/access_policies.html);
- Then proceed with the new user creation setup.

Upon generation, the IAM role will provide a public **Access Key ID** and **Secret Access Key**. Download this file for reference later.

:::warning Warning
The **Secret Access Key** provides access to your AWS user account. It should be treated like any other secret credentials. Specifically, it should never be checked into source control.
:::

> By default, Nebari will try to use the credentials associated with the current AWS infrastructure/environment for authentication. Please keep in mind that Nebari will only use these credentials to create the first roles and stricter permissions for Nebari's internal components. Refer to [Conceptual guides] for more information on how Nebari's components are secured.

Reviewer: > By default, Nebari will try to use the credentials associated with the current AWS infrastructure/environment for authentication.

As opposed to what other credentials?

> Please keep in mind that Nebari will only use these credentials to create the first roles and stricter permissions for Nebari's internal components.

Is this meant to emphasize that Nebari is only going to use the *administrator level* credentials to create roles with lesser permissions, and then those lower-permission roles will be used to operate Nebari's internal components?

----

There's an incomplete link here on `[Conceptual guides]` -- looks like the page just may not exist yet?

Will there also be a conceptual guide as to how Nebari's internal components are laid out, and not just how they're secured?

Author: @bskinn let me know what you think, any ideas on how I can re-write that part?

Reviewer 2: I think this is ok - may be worth linking to the AWS (and equivalent for other providers) as to where AWS looks for credentials: https://docs.aws.amazon.com/sdkref/latest/guide/file-location.html

Provide authentication credentials to Nebari by setting the following environment variables:

```bash
export AWS_ACCESS_KEY_ID="Access Key ID"
export AWS_SECRET_ACCESS_KEY="Secret Access Key"
```

> If you are using an already existing IAM user or want to use a different par of access keys, please refer to [Managing access keys for IAM users](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_credentials_access-keys.html#Using_CreateAccessKey) for detailed information on how to manage your IAM user's access keys.

Reviewer: I think this should be relocated to be the second paragraph in this "Authentication" section -- it should be somewhere that someone who already has an IAM user and keys will find it quickly, instead of having to read to the bottom of the section.

Is there any reason other than wanting to use an existing IAM user that would motivate someone to use a different pair of access keys? If not, then "If you are using an existing IAM user or want to use a different pair of access keys" is redundant.

```suggestion
If you are using an existing IAM user or want to use a different pair of access keys, please refer to [Managing access keys for IAM users](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_credentials_access-keys.html#Using_CreateAccessKey) for detailed information on how to manage your IAM user's access keys.
```

:::tip

> This variables applies only to your current shell session. If you want the variable to apply to future shell sessions, set the variable in your shell startup file, for example in the `~/.bashrc` or `~/.profile` file.

Reviewer: ```suggestion
These environment variables will apply only to your current shell session. If you want the variables to apply to future shell sessions also, set the variables in your shell startup file (for example, in `~/.bashrc` or `~/.profile` for the bash shell).
```

:::

:::note

> The following steps assumes you have completed the [Install Nebari](/started/installing-nebari) section, has confirmed that `nebari` is successfully installed in your environment and opted for **AWS** as your cloud provider. as well as has already configured the `nebari` environment variables. If you had any issues during the installation, please visit our getting started [troubleshooting](/started/troubleshooting) section for extra advice.

Reviewer: ```suggestion
The steps in the following sections assume you have (i) completed the [Install Nebari](/started/installing-nebari) section, (ii) confirmed that `nebari` is successfully installed in your environment, (iii) opted for **AWS** as your cloud provider, and (iv) already configured the `nebari` environment variables. If you had any issues during the installation, please visit the "Getting Started" section of our [troubleshooting page](/started/troubleshooting) for further guidance.
```

:::

## Nebari Initialize

> Great, you’ve gone through the [Nebari Installation](/started/installing-nebari.md) and [Authentication setup](#authentication) steps, and have ensured that all the necessary environment variables have been properly set. It is time to initialize and deploy Nebari!

Reviewer: ```suggestion
Great, you’ve gone through the [Nebari Installation](/started/installing-nebari.md) and [authentication setup](#authentication) steps, and have ensured that all the necessary environment variables have been properly set. It is time to initialize and deploy Nebari!
```

> In your terminal, we advise you to start by creating a new project folder. Here, we will name the new folder as `nebari-aws`:

Reviewer: ```suggestion
In your terminal, start by creating a new project folder. For this demonstration, we will name the new folder `nebari-aws`:
```

```bash
mkdir nebari-aws && cd nebari-aws
```
When you first initialize Nebari, you will be creating a `nebari-config.yaml` that contains a collection of preferences and settings for your deployment. The command bellow will generate a basic config file with an infrastructure based on **AWS**, named **projectname**, where the domain will be **domain** and the authetication mode set to **password**.

```bash
nebari init aws --project projectname --domain domain --auth-provider password
```
:::note Note

> You will be prompted to enter values for some of the choices above if they are omitted as command line arguments (for example project name and domain)

Reviewer: ```suggestion
You will be prompted to enter values for some of the choices above if they are absent from the command line arguments (for example, project name and domain)
```

:::

Once `nebari init` is executed, you should then be able to see the following output:
```bash
Securely generated default random password=*** for Keycloak root user
stored at path=/tmp/NEBARI_DEFAULT_PASSWORD
INFO:botocore.credentials:Found credentials in environment variables.
```
:::tip

> MacOS generates a programmatic directory stored in `/private/var` and defines the `$TMPDIR` environment variable for locating the system temporary folder. We advise the user to look for the `NEBARI_DEFAULT_PASSWORD` file in the following  `/var/folders/xx/xxxxx/T` path.

Author: What about:

The main `temp` folder on a MacOS system can be found by inspecting the value of `$TMPDIR`. This folder and its files are not meant to be user-facing and will present you with a seemingly random directory path similar to `/var/folders/xx/xxxxx/T`

:::

> You can see that Nebari is generating a random password for the root user of Keycloak. This password is stored in a temporary file and will be used to authenticate to the Keycloak server once Nebari's infrastructure is fully deployed.

Reviewer: What happens to this password? Why is it stored in a temporary file? Doesn't that run the risk of locking us out of our Nebari deployment? 

If the plan is to create one or more non-root administrator users as part of standing up the Nebari instance, it would be good to mention that here. Even if this Ubuntu-like superuser workflow is common in some circles, I still think it would be a good idea to be explicit here about how this will be handled.

Suggested change, assuming my understanding is correct:

```suggestion
You can see that Nebari is generating a random password for the root user of Keycloak. This password is stored in a temporary file and will be used to authenticate to the Keycloak server once Nebari's infrastructure is fully deployed, in order to create the first user accounts for administrator(s).
```

Author: > What happens to this password? Why is it stored in a temporary file? Doesn't that run the risk of locking us out of our Nebari deployment?

The same password is stored in the `nebari-config` file, so you can always use that even when the credential is missing in the env. 

> If the plan is to create one or more non-root administrator users as part of standing up the Nebari instance, it would be good to mention that here. Even if this Ubuntu-like superuser workflow is common in some circles, I still think it would be a good idea to be explicit here about how this will be handled.

The idea is to have only `root` as the administrator of keycloak, and we advise the user to change the password. But all this is out of the scope of installation and deployment, this point is addressed in the [Keycloak docs](https://docs.qhub.dev/en/stable/source/installation/login.html?#change-keycloak-root-password) of Nebari (I linked the current ones).

> The generated `nebari-config.yaml` is the configuration file that will determine how the cloud infrastructure and Nebari is built and deployed in the next step. But at this point it’s just a text file. You could edit it manually if you are unhappy with the choices, or delete it and start over again.

Reviewer: ```suggestion
The generated `nebari-config.yaml` is the configuration file that will determine how the cloud infrastructure and Nebari is built and deployed in the next step. Since it is a simple text file, you can edit it manually if you are unhappy with the choices you made during initialization, or delete it and start over again by re-running `nebari init`.
```

For additional information about the `nebari-config.yaml` file and extra flags that allow you to configure the initialization , see the [Understanding the nebari-config.yaml file](/tutorials/overview.md) documentation.

## Deploying Nebari

> Finally, with the `nebari-config.yaml` created, Nebari can be deployed for the first time:

Reviewer: ```suggestion
With the `nebari-config.yaml` configuration file now created, Nebari can be deployed for the first time:
```

```bash
nebari deploy -c nebari-config.yaml
```

> The terminal will prompt you to press `[enter]` to check auth credentials, which were added in the previous step by the initialization command. Once authenticated, the infrastructure deployment process will start and might take around a couple minutes to complete.

Reviewer: ```suggestion
The terminal will prompt you to press `[enter]` to check the authentication credentials that were added as part of the preceding `nebari init` command. Once Nebari is authenticated, it will start its infrastructure deployment process, which will take up to a couple of minutes to complete.
```

> Once you reach stage `04-kubernetes-ingress` you will be prompted to set the **A/CNAME** records manually for your registered domain name. Please follow the instructions in the [Nebari DNS](/how-tos/domain-registry.md) section for more information regarding the domain `A/CNAME` records and how to automatically generate them.

Reviewer: Will users always be wanting to *automatically* generate their `A/CNAME` records?

```suggestion
Once you reach the `04-kubernetes-ingress` stage of the deploy process, you will be prompted to set the **A/CNAME** records manually for your registered domain name. Please follow the instructions in the [Nebari DNS](/how-tos/domain-registry.md) section for more information regarding the domain `A/CNAME` records and how to automatically generate them.
```

If the deployment is successful, you will see the following output:
```bash
[terraform]: Nebari deployed successfully
Services:
 - argo-workflows -> https://projectname.domain/argo/
 - conda_store -> https://projectname.domain/conda-store/
 - dask_gateway -> https://projectname.domain/gateway/
 - jupyterhub -> https://projectname.domain/
 - keycloak -> https://projectname.domain/auth/
 - monitoring -> https://projectname.domain/monitoring/

> Kubernetes kubeconfig located at file:///tmp/NEBARI_KUBECONFIG

Author: @bskinn yes it will be cleaned up, but nebari re-generates the file for every run, so we don't have any problems. I agree that this question is important, but this should be added to troubleshooting as `Can't find NEBARI_KUBECONFIG file on /tmp, what should I do?` or a similar question, right?

> Kubernetes kubeconfig located at file:///tmp/NEBARI_KUBECONFIG

Reviewer 3: This issue will be addressed in [here](https://github.com/nebari-dev/nebari/issues/83#issuecomment-1171641391)

:::

You can see that Nebari is generating a random password for the root user of Keycloak. This password is stored in a temporary file and will be used to authenticate to the Keycloak server once Nebari's infrastructure is fully deployed, in order to create the first user accounts for administrator(s).

The generated `nebari-config.yaml` is the configuration file that will determine how the cloud infrastructure and Nebari is built and deployed in the next step. Since it is a simple text file, you can edit it manually if you are unhappy with the choices you made during initialization, or delete it and start over again by re-running `nebari init`.

For additional information about the `nebari-config.yaml` file and extra flags that allow you to configure the initialization process, see the [Understanding the nebari-config.yaml file](/tutorials/overview.md) documentation.

## Deploying Nebari

With the `nebari-config.yaml` configuration file now created, Nebari can be deployed for the first time:

```bash
nebari deploy -c nebari-config.yaml

> ```

Author: ?

---
id: nebari-aws
title: How to deploy Nebari on AWS
---

# How to deploy Nebari on AWS

A basic overview to deploy Nebari on AWS

:::note
The following steps assumes you have just completed the [Install Nebari](/started/installing-nebari) section, has confirmed that `nebari` is successfully installed in your environment and opted for **AWS** as your cloud provider. If you had any issues in the prior steps please visit our getting started [troubleshooting](/started/troubleshooting) section for extra advice.
:::

## Authentication

> By default, Nebari will try to use the credentials associated with the current AWS infrastructure/environment for authentication.

Author: I've tried to cover that up now, in the authentication sections.

- Please see these instructions for [creating an IAM](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_create.html) role with administrator permissions.
- Upon generation, the IAM role will provide a public **access key ID** and a **secret key** which will need to be added to the environment variables.

To define the environment variables paste the commands below with your respective keys.

```bash
export AWS_ACCESS_KEY_ID="HAKUNAMATATA"
export AWS_SECRET_ACCESS_KEY="iNtheJUng1etheMightyJUNgleTHEl10N51eEpsT0n1ghy;"

> ```

Reviewer 3: if I run these commands in a terminal, they become environment variable in that terminal window (i.e. not permanent). Will I need to use these again? Should they be in a permanent place (like .bashrc, etc). Are they only used for the "installation" steps?

> ```

Author: That's now included, thanks @kcpevey
